test/src/unit-capi-dynamic.cc

```diff
@@ -0,0 +1,841 @@
+    buffer_sizes.emplace_back(buffer_size);
+  }
+
+  // Create the dimension write buffers.
+  // NOTE: if there is more than 1 dimension, the buffer shall be the previous
+  // buffer multiplied by the current range
+  vector<uint64_t*> d_write_buffers;
+  auto buff_size_iter = buffer_sizes.begin();
+  while (buff_size_iter != buffer_sizes.end()) {
+    uint64_t* write_buffer;
+    if (buff_size_iter == buffer_sizes.begin()) {
+      write_buffer = (uint64_t*)malloc(*buff_size_iter * sizeof(uint64_t*));
+    } else {
+      write_buffer = (uint64_t*)malloc(
+          *buff_size_iter * (*buff_size_iter - 1) * sizeof(uint64_t*));
+    }
+    d_write_buffers.emplace_back(write_buffer);
+    buff_size_iter = std::next(buff_size_iter, 1);
+  }
+
+  // Define the write query buffers for "a".
+  vector<test_query_buffer_t> write_query_buffers;
+  uint64_t a_write_buffer[1] = {0};
+  uint64_t a_write_buffer_size = sizeof(a_write_buffer);
+  uint8_t a_write_buffer_validity[1] = {static_cast<uint8_t>(rand() % 2)};
+  uint64_t a_write_buffer_validity_size = sizeof(a_write_buffer_validity);
+  int a_write_buffer_var[2] = {0, 1};
+  uint64_t a_write_buffer_var_size = sizeof(a_write_buffer_var);
+
+  // Validity buffer is set only when the attribute is nullable.
+  // Variable buffer is set only when the attribute is var-sized.
+  if (test_attr.nullable_) {
+    if (test_attr.cell_val_num_ == TILEDB_VAR_NUM) {
+      write_query_buffers.emplace_back(
+          test_attr.name_,
+          a_write_buffer,
+          &a_write_buffer_size,
+          a_write_buffer_var,
+          &a_write_buffer_var_size,
+          a_write_buffer_validity,
+          &a_write_buffer_validity_size);
+    } else {
+      write_query_buffers.emplace_back(
+          test_attr.name_,
+          a_write_buffer,
+          &a_write_buffer_size,
+          nullptr,
+          nullptr,
+          a_write_buffer_validity,
+          &a_write_buffer_validity_size);
+    }
+
+  } else {
+    if (test_attr.cell_val_num_ == TILEDB_VAR_NUM) {
+      write_query_buffers.emplace_back(
+          test_attr.name_,
+          a_write_buffer,
+          &a_write_buffer_size,
+          a_write_buffer_var,
+          &a_write_buffer_var_size,
+          nullptr,
+          nullptr);
+    } else {
+      write_query_buffers.emplace_back(
+          test_attr.name_,
+          a_write_buffer,
+          &a_write_buffer_size,
+          nullptr,
+          nullptr,
+          nullptr,
+          nullptr);
+    }
+  }
+
+  // Define dimension query write vectors for either sparse arrays
+  // or dense arrays with an unordered write order.
+  if (array_type == TILEDB_SPARSE || write_order == TILEDB_UNORDERED) {
+    // Fill the write buffers with values.
+    auto dims_iter = test_dims.begin();
+    for (auto write_buff_iter = d_write_buffers.begin();
+         write_buff_iter != d_write_buffers.end();
+         ++write_buff_iter) {
+      for (uint64_t i = 0; i < sizeof(write_buff_iter); i++) {
```

Comment:
I believe you want `write_buff_iter->size()` instead of `sizeof(write_buff_iter)`

Reply:
Don't we want the size of each individual element in the vector here so we can fill each buffer accordingly? Perhaps I need to divide `sizeof(write_buff_iter) / sizeof(uint64_t)` to get the true value I want. I believe doing it the way you're suggesting will give me the error `Expression must have pointer-to-class type`.

Reply:
I see -- these elements are `uint64_t*`, I wrote this thinking that they were of type `vector`. My point still stands: `sizeof(write_buff_iter)` is returning the size of the `std::vector<uint64_t>::iterator` class, it is not returning the length of the buffer.

You could get this size from `buffer_sizes[i]`.